mmtrack/models/sot/stark.py

```diff
@@ -0,0 +1,247 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from collections import defaultdict
+
+import torch
+import torch.nn.functional as F
+from mmdet.models.builder import build_backbone, build_head, build_neck
+from torch.nn.modules.batchnorm import _BatchNorm
+from torch.nn.modules.conv import _ConvNd
+
+from ..builder import MODELS
+from .base import BaseSingleObjectTracker
+
+
+@MODELS.register_module()
+class Stark(BaseSingleObjectTracker):
+    """STARK: Learning Spatio-Temporal Transformer for Visual Tracking.
+
+    This single object tracker is the implementation of `STARk
+    <https://arxiv.org/abs/2103.17154>`_.
+
+    Args:
+        backbone (dict): the configuration of backbone network.
+        neck (dict, optional): the configuration of neck network.
+            Defaults to None.
+        head (dict, optional): the configuration of head network.
+            Defaults to None.
+        init_cfg (dict, optional): the configuration of initialization.
+            Defaults to None.
+        frozen_modules (str | list | tuple, optional): the names of frozen
+            modules. Defaults to None.
+        train_cfg (dict, optional): the configuratioin of train.
+            Defaults to None.
+        test_cfg (dict, optional): the configuration of test.
+            Defaults to None.
+    """
+
+    def __init__(self,
+                 backbone,
+                 neck=None,
+                 head=None,
+                 init_cfg=None,
+                 frozen_modules=None,
+                 train_cfg=None,
+                 test_cfg=None):
+        super(Stark, self).__init__(init_cfg)
+        self.backbone = build_backbone(backbone)
+        self.neck = build_neck(neck)
+        self.head = build_head(head)
+
+        self.test_cfg = test_cfg
+        self.train_cfg = train_cfg
+
+        # Set the update interval
+        self.update_intervals = self.test_cfg['update_intervals']
+        self.num_extra_template = len(self.update_intervals)
+
+        self.debug = False
+        if frozen_modules is not None:
+            self.freeze_module(frozen_modules)
+        # We only train classification head in stage-2. The transformer and
+        # bbox head are not trained in stage-2.
+        if self.head.run_cls_head and not self.head.run_bbox_head:
+            for name, module in self.head.named_parameters():
```

Comment:
use `self.freeze_module` to freeze modules.
And `self.head.run_cls_head` and `self.head.run_bbox_head` can be ommited.

Reply:
The `backbone` and `neck` in the configs can be freezed by `self.freeze_module`. The `head ` module contains 'transformer'', `bbox_head` and `cls_head`, we can not directly used `self.freeze_module` to only freeze the `transformer` and `bbox_head` except `cls_head`.

Reply:
We can add a `freeeze_module` in the head.
In summary, we need to delete `run_bbox_head` and `run_cls_head`, since it is hard code.
How to freeze the module in the model should be configed flexiblely in the config.